README.md
# SHST

[![tips2](https://img.shields.io/badge/-API-%234C98F7.svg?style=for-the-badge&logo=monogram&logoColor=white)](https://github.com/WindrunnerMax/SW/tree/master) 
[![tips1](https://img.shields.io/badge/-更新日志-%234C98F7.svg?style=for-the-badge&logo=azure-pipelines&logoColor=white)](https://github.com/WindrunnerMax/SHST/blob/SDUST/ChangeLog.md) 
[![tips3](https://img.shields.io/badge/-山科小站-%234C98F7.svg?style=for-the-badge&logo=marketo&logoColor=white)](https://windrunner_max.gitee.io/imgpath/SHST/Static/SHST-WX.jpg) 

```
SHST/Python 目录下为Python版本的API
SHST/PHP 目录下为PHP版本的API
SHST/Java 目录下为Java版本的API 
SHST/SHST-UNI 目录下为小程序项目[山科小站--小程序]，山东科技大学校园服务平台，已上线微信小程序与QQ小程序 
SHST/SHST-WEL 目录下为小程序项目[山科小站--迎新专版]，提供校内的相关信息，迎新专用，已上线微信小程序与QQ小程序   
SHST/SHST-WEX 目录下为NVUE/WEEX项目[山科小站--APP]，采用原生渲染，作为UNIAPP纯NVUE/WEEX项目开发，已上架酷安应用市场
```


## API 

### Python

```python
# Python/SW.py

# 配置信息
account = ""                              # 账号
password = ""                             # 密码
url = "http://jwgl.sdust.edu.cn/app.do"   # ${学校教务系统}/app.do

# 代码末尾注释的方法 取消注释测试执行
Q.get_student_info()                      # 获取学生信息
Q.get_current_time()                      # 获取学年信息
Q.get_class_info()                        # 当前周次课表
Q.get_class_info(3)                       # 指定周次课表
Q.get_classroom_info("0102")              # 空教室查询 "allday"：全天 "am"：上午 "pm"：下午 "night"：晚上 "0102":1.2节空教室 "0304":3.4节空教室
Q.get_grade_info("2018-2019-1")           # 成绩查询 # 无参数查询全部成绩
Q.get_exam_info()                         # 获取考试信息
```

### PHP
```php
// PHP/Main.php

// 配置信息
$accountSW = "";                            // 账号
$passwordSW = "";                           // 密码
$urlSW = "http://jwgl.sdust.edu.cn/app.do"; // ${学校教务系统}/app.do

// 代码末尾注释的方法 取消注释测试执行
$Q -> getStudentInfo();                     // 获取学生信息
$Q -> getCurrentTime();                     // 获取学年信息
$Q -> getTable();                           // 当前周次课表
$Q -> getTable(3);                          // 指定周次课表
$Q -> getGrade("2018-2019-2");              // 成绩查询 // 无参数查询全部成绩
$Q -> getClassroom("0102");                 // 空教室查询 "allday"：全天 "am"：上午 "pm"：下午 "night"：晚上 "0102":1.2节空教室 "0304":3.4节空教室
$Q -> getExam();                            // 获取考试信息
```

### Java
```java
// Java/MainSw.java

// 配置信息
private String account = "";                            // 账号
private String password = "";                           // 密码
private String url = "http://jwgl.sdust.edu.cn/app.do"; // ${学校教务系统}/app.do

// 代码末尾注释的方法 取消注释测试执行
Q.getStudentInfo().exec();                              // 获取学生信息
Q.getCurrentTime().exec();                              // 获取学年信息
Q.getTable().exec();                                    // 当前周次课表
Q.getTable().setWeek("3").exec();                       // 指定周次课表
Q.getGrade().exec();                                    // 查询全部成绩
Q.getGrade().setTerm("2018-2019-2").exec();             // 查询指定学期成绩
Q.getClassroom("0102").exec();                          // 空教室查询 "allday"：全天 "am"：上午 "pm"：下午 "night"：晚上 "0102":1.2节空教室 "0304":3.4节空教室
Q.getExamInfo().exec();                                 // 获取考试信息
```

### Notice

```php
/***
注意：
1. 由于强智版本不尽相同，返回的数据字段会有所差别，例如我们学校返回的是flag字段标记登陆成功，而有学校会返回success字段标记登陆成功
2. 数据接口全部抓取智校园App而来，可以使用Fiddler等抓包工具自行尝试抓包，注意安卓7及以上不会认同用户自定义证书，可以使用root将证书安装为系统证书或使用其他的辅助工具尝试抓包
3. 虽然数据接口由智校园App得来，但这并不意味着只有学校支持智校园才能使用数据接口，强智教务系统的接口一般是默认开放的，当然系统管理员可以手动关闭，而智校园的使用是需要强智公司授权的，也就是说虽然学校不能用智校园，但是完全有可能开放接口
4. 目前发现有的教务系统不能直接查询全部成绩，这个接口的使用请自行验证，按学期查询成绩的接口使用目前并未发现问题
5. 如果接口无法使用，可以尝试直接识别验证码爬取教务系统，验证码识别请看 https://github.com/WindrunnerMax/SWVerifyCode 此仓库，提供了使用 Python、PHP、Java、JavaScript 识别验证码的示例
6. 对于请求的URL，是直接使用protocol://hostname[:port]/app.do，并不是教师学生端的URL再拼接app.do，具体可以访问http://app.qzdatasoft.com:9876/qzkjapp/phone/provinceData查阅，此外有些使用ASPX的强智系统的开放接口为${学校教务系统}/app/app.ashx
***/
```

  
## 山科小站

![山科小站二维码](https://cdn.jsdelivr.net/gh/SHST-SDUST/SHST-UNI/vector/resources/exhibition/shst-wx.jpg)

### 配置相关 

```javascript
// SHST-UNI/App.vue

url            // 后台请求域名
tips           // 公告本地标识
header         // 请求头信息
openid         // OPENID信息
version        // 版本号
curTerm        // 当前学期
initData       // 初始化数据信息
userFlag       // 用户登录状态
colorList      // 颜色方案
curTermStart   // 开学日期
ajax()         // 网络请求封装 回调
toast()        // 弹窗提示
extend()       // 深拷贝与浅拷贝
request()      // 网络请求封装 Promise
```


### 目录结构  

[关于UNIAPP重构以及类的封装文档](https://blog.touchczy.top/#/MiniProgram/%E5%B1%B1%E7%A7%91%E5%B0%8F%E7%AB%99%E5%B0%8F%E7%A8%8B%E5%BA%8F)

```
SHST-UNI                              // 山科小站总目录
    ├── components                    // 组件封装
    │   ├── headslot.vue              // 带solt的标题布局
    │   ├── layout.vue                // 卡片式布局
    │   ├── list.vue                  // 展示用list布局
    │   ├── sentence.vue              // 每日一句封装
    │   └── weather.vue               // 天气封装
    ├── modules                       // 模块化封装
    │   ├── cookies.js                // Cookies操作
    │   ├── copy.js                   // 深浅拷贝
    │   ├── datetime.js               // 时间日期操作
    │   ├── event-bus.js              // 事件总线
    │   ├── global-data.js            // 全局变量
    │   ├── loading.js                // 加载提示
    │   ├── operate-limit.js          // 防抖与节流
    │   ├── regex.js                  // 正则匹配
    │   ├── request.js                // 网络请求
    │   ├── toast.js                  // 消息提示
    │   └── update.js                 // 自动更新 
    ├── pages                         // 页面
    │   ├── Ext                       // 拓展组
    │   ├── Home                      // Tabbar、辅助组
    │   ├── Lib                       // 图书馆功能组
    │   ├── Sdust                     // 科大组
    │   ├── Study                     // 学习组
    │   └── User                      // 用户组
    ├── static                        // 静态资源
    │   ├── camptour                  // 校园导览静态资源
    │   └── img                       // 图标等静态资源
    ├── unpackage                     // 打包文件
    ├── utils                         // 辅助功能
    │   ├── amap-wx.js                // 高德地图SDK
    │   └── md5.js                    // MD5引入
    ├── vector                        // 部署封装
    │   ├── resources                 // 资源文件
    │   │   ├── camptour              // 校园导览配置文件
    │   │   ├── asse.mini.wxss        // 公共样式库
    │   │   └── iconfont.wxss         // 字体图标
    │   ├── dispose.js                // 部署小程序
    │   └── pubFct.js                 // 公有方法
    ├── App.vue                       // App全局样式以及监听
    ├── main.js                       // 挂载App，Vue初始化入口文件
    ├── manifest.json                 // 配置Uniapp打包等信息
    ├── pages.json                    // 路由
    └── uni.scss                      // 内置的常用样式变量
```

### 小程序  
![山科小站宣传图](https://cdn.jsdelivr.net/gh/SHST-SDUST/SHST-UNI/vector/resources/exhibition/show.jpg)



ChangeLog.md
## Changelog

```
SDUST
```


### 3.5.7 (2021-06-15) [b32db09]
1. 修正一些变量定义方式
2. 修复无空教室时的异常

### 3.5.6 (2021-05-21) [c1d1607]
1. 增加泰安校区空教室查询
2. 调整字体大小等细节处理

### 3.5.5 (2021-03-23) [ade1cb6]
1. 处理课表异常报错
2. 增加实时日志调用
3. 优化校园卡样式结构

### 3.5.4 (2021-03-09) [ea8002a]
1. 增加新鲜事功能
2. 增加scss样式库
3. 调整部分提示信息
4. 修复按钮样式的兼容问题
5. 图书馆增加图片与条目调整

### 3.5.3 (2021-01-19) [685e6f9]
1. 修复图床图片资源问题
2. 修复校园导览部分细节
3. 缓存控制对象全面部署
4. 添加Mixin全局静态横切
5. 修复IOS对于Date对象兼容问题

### 3.5.2 (2021-01-17) [752fb97]
1. 修正文字的描述信息
2. 修复每日一句接口问题
3. 增加缓存控制对象封装
4. 广告组件加入后端控制展示

### 3.5.1 (2020-12-08) [e5c2961]
1. 修复校历抛出的异常
2. 修复自定义课表保存逻辑问题
3. 添加事件总线一次性触发订阅
4. 修复初始化失败后无法继续操作的问题

### 3.5.0 (2020-11-22) [717ef8e]
1. 解决初始化状态问题
2. 修复校历假期颜色问题
3. 自定义添加课表项目功能

### 3.4.1 (2020-10-25) [1ef3d5d]
1. 天气组件样式修正
2. 加入公告仅弹出情况
3. 处理课表新的数据结构
4. 去除教室课表与蹭课查询广告位

### 3.4.0 (2020-10-24) [bbd9c16]
1. 加入轮播图
2. 天气样式修改
3. 加入弹出式公告
4. 修复校内公告图片显示问题
5. 适配济南校区空教室查询功能

### 3.3.6 (2020-10-07) [092a6fa]
1. 增加校内公告功能
2. 修复赞赏列表的点击样式问题

### 3.3.5 (2020-10-06) [52e3c52]
1. 广告组件以及兼容处理
2. 修正共享课表数据拉取失败问题
3. 修正登录页记住密码问题
4. 修正校园卡加载失败的提示

### 3.3.4 (2020-09-19) [6d83e3f]
1. 简单语句内联处理
2. 修复测试账号问题
3. 按教室查询课程功能
4. 按课程名与教师名蹭课查询功能

### 3.3.3 (2020-09-12) [e552f70]
1. 处理异常数据触发的信息提示 
2. 修复网络请求地址异常的问题

### 3.3.2 (2020-09-10) [729b6f5]
1. 修复校历开学日期问题
2. 修复共享课表刷新问题
3. 修复教室查询请求失败问题
4. 修复首页公告无法打开问题
5. 修复绑定教务系统无跳转问题

### 3.3.1 (2020-09-03) [2decef8]
1. 解决样式冲突问题
2. 修正公共样式文件问题
3. 解决同时段课程信息覆盖问题
4. 完成迎新版与App版的代码重构

### 3.3.0 (2020-08-31) [2951ea9]
1. 修正变量命名规范
2. 增加防抖与节流模块
3. 完善request模块封装
4. 修正cookies模块功能
5. 修正组件订阅登录事件
6. 冗余处理，压缩代码体积
7. 完善ES6标准模块导入导出
8. 配合后端服务统一判定标识

### 3.2.2 (2020-04-16) [02078ab]
1. Promise与async/await
2. 增加测试账号的功能
3. 重写图书馆爬虫方案
4. 增加ES6标准模块导出
5. 自定义课表配色功能

### 3.2.1 (2020-03-25) [c4241dc]
1. 更新校历UI
2. 更新课表配色方案

### 3.2.0 (2020-02-08) [7d18649]
1. 修复link样式问题
2. 修改静态资源路径
3. App增加科大公告功能
4. App页面填充算法优化

### 3.1.3 (2020-02-05) [4d74161]
1. 小程序课表时间修正
2. App布局Flex样式修正

### 3.1.2 (2020-01-28) [064c506]
1. App Nvue/Weex Finish

### 3.1.1 (2019-12-30) [c544ff4]
1. 山科小站App完成

### 3.1.0 (2019-12-28) [e2cf6a7]
1. 山科小站迎新重构完成
2. 山科小站教务重构完成

### 3.0.0 (2019-08-31) [ccff764]
1. 赞赏列表
2. 课表全面缓存
3. 校园导览修正

### 2.9.3 (2019-08-26) [d7c93b0]
1. 修复图书馆借阅超期显示问题

### 2.9.2 (2019-08-20) [808fc0f]
1. 小程序接入公众号

### 2.9.1 (2019-08-16) [166793e]
1. 公众号消息功能

### 2.8.3 (2019-08-15) [b83401a]
1. 跳转迎新版
2. 图标数据修正

### 2.8.2 (2019-08-13) [b1a81b4]
1. 公众号组件
2. Onlunch Event

### 2.8.1 (2019-08-07) [cb8dbac]
1. 订阅发布模式

### 2.7.5 (2019-08-06) [2796d27]
1. 待办事项缓存

### 2.7.4 (2019-08-03) [1404bc3]
1. 字体图标iconfont
2. 增加赞赏功能

### 2.7.3 (2019-08-02) [1f8eae3]
1. Redis存储
2. 增加后台管理功能
3. 修复校园卡与图书馆问题

### 2.7.2 (2019-07-30) [0167719]
1. 校园卡查询功能

### 2.7.1 (2019-07-27) [9618313]
1. 校园导览功能完成
2. 动态公告功能
3. HTTP请求类请求头处理，响应头处理
4. 小程序目录结构处理
5. 后端冗余处理

### 2.6.2 (2019-07-23) [f429f44]
1. 校园导览优化定位逻辑
2. 解决注销切换用户问题

### 2.6.1 (2019-07-21) [d69a6f6]
1. 校园导览功能上线

### 2.5.6 (2019-07-17) [316818d]
1. API更新
2. 解决分享引起的问题

### 2.5.5 (2019-07-16) [0c6cd5d]
1. QQ小程序 山科小站--迎新专版

### 2.5.4 (2019-07-15) [f22f000]
1. 添加自定义组件
2. About界面重构

### 2.5.3 (2019-07-10) [d25a95e]
1. 添加公告红点
2. 重构日历功能
3. API兼容处理(暂)

### 2.5.2 (2019-07-07) [34c257c]
1. 更新Python版本API
2. 添加PHP版本API
3. 添加Java版本API

### 2.5.1 (2019-07-06) [36fcf0f]
1. 统一版本号
2. 启用备用域名
3. 版本更新功能
4. Application层
5. 优化Http请求类结构
6. 修复App的foreach问题

### 2.4.5 (2019-07-02) [c0d33f6]
1. 缓存用户信息
2. 封装的时间日期修复，拓展Date原型addDate方法
3. 更改结构，加入dispose.js公共数据处理方法

### 2.4.4 (2019-06-27) [0f65966]
1. 因后台逻辑问题共享课表只显示周次
2. 处理Toast执行顺序问题
3. 考试安排功能

### 2.4.3 (2019-06-25) [a64d565]
1. 部分加载提示载入封装网络请求的Complete中
2. 空教室排序
3. 登陆时显示密码开关

### 2.4.2 (2019-06-24) [d813066]
1. 优化APP请求逻辑
2. 修复版本号问题

### 2.4.1 (2019-06-22) [85a5d31]
1. 登录逻辑更改
2. 优化封装网络请求

### 2.3.2 (2019-06-21) [59cecbd]
1. 查询教室功能优化

### 2.3.1 (2019-06-19) [a0b96cf]
1. 近7天空教室查询功能
2. 修复部分冗余

### 2.2.5 (2019-06-18) [e38c011]
1. 目录结构整理
2. 修复时间与登录BUG

### 2.2.4 (2019-06-16) [3f97a7b]
1. 使用Storage存储部分数据

### 2.2.3 (2019-06-15) [3c09247]
1. 更新前后端逻辑，减少对强智的请求

### 2.2.2 (2019-06-13) [74ec7d2]
1. 代码目录结构
2. user界面重构
3. 修复一些登录的问题

### 2.2.1 (2019-06-11) [ed5f91b]
1. 减少服务器负载，本地处理数据

### 2.1.2 (2019-06-09) [1b33ddf]
1. 待办安排功能

### 2.1.1 (2019-06-08) [f141520]
1. 图标颜色问题
2. 修复主页用户未授权的问题
3. 增加放假安排功能

### 2.0.5 (2019-06-07) [c59e024]
1. 修复GET请求编码问题(小程序+APP)

### 2.0.4 (2019-05-26) [d00a5b0]
1. 微信小程序提供游客模式，完善校历和嵙地图功能

### 2.0.3 (2019-05-23)[ace5f30]
1. 微信小程序完成教室，成绩，共享课表，图书馆，借阅，地图，校历，链接

### 2.0.2 (2019-05-22)[3353a8a]
1. 微信小程序完成课表、用户

### 2.0.1 (2019-05-21) [3f51c42]
1. 微信小程序开发(WXSA)
2. 完成微信小程序登录、今日课程、天气

### 1.3.9 (2019-05-18) [B0aaa8b]
1. 增加近三天天气功能，点击图标查看今日天气状态

### 1.3.8 (2019-05-17) [1d4f1e3]
1. 优化颜色算法
2. 修复课表因课重复而覆盖问题

### 1.3.7 (2019-05-11) [56522c5]
1. 移除部分功能

### 1.3.6 (2019-05-09) [3f9c6da]
1. 关闭伪静态(增加index.php)

### 1.3.5 (2019-05-05) [a12bf8c]
1. 增加情侣课表功能
2. 折叠图书馆功能块

### 1.3.4 (2019-05-04) [821b4b2]
1. 修复图书检索为GET方式
2. 课表优化代码，加入背景颜色

### 1.3.3 (2019-05-03) [e3d83f2]
1. 增加链接分享功能

### 1.3.2 (2019-05-01) [7b4b712]
1. 整理代码结构
2. 完善图书馆检索与借阅查询功能
3. 学校图书馆对外网开放时间大约在 7:00 - 20:00 ，图书功能在此时间段正常使用

### 1.3.1 (2019-04-30) [642c52f]
1. 优化部分代码结构
2. 建立SDUST分支(嵙分支，后期commit基本于此分支)
3. 加入图书馆检索图书功能
4. 加入借阅信息查询功能

### 1.2.2 (2019-04-28) [cc2f702]
1. 优化代码结构，建立配置静态类

### 1.2.1 (2019-04-20) [80c3102]
1. 内网穿透，恢复使用

### 1.2.0 (2019-04-11) [9903ae4] 
1. 增加兼容模式(兼容模式是非异步的，可能会稍有卡顿)
2. 增加 更新日志 与 公告
3. 使用localStorage作为本地存储

### 1.1.1 (2019-04-10) [02b4603] 
1. 增加Github入口

### 1.1.0 (2019-03-11) [f041eb2] 
1. 更改数据获取方式为异步
2. 增加成绩查询与绩点计算功能

### 1.0.0 (2019-03-05) [-------] 
1. 最早是因为智校园无法使用，我无法查询自习室，在J7爬了四层楼未找到自习室，遂回宿舍封装了此Web
2. 由于最初版本仅为自用，并未使用Git管理代码，所以无commit
3. 服务端未保存任何信息，账号与密码全部以cookies保存在本地


Python/SW.py
#!/usr/bin/python 
# -*- coding: utf-8 -*-

import time
import requests
import re
import os
import json
import datetime

# 强智教务管理系统
########################################
account = ""
password = ""
url = "http://jwgl.sdust.edu.cn/app.do" 
########################################


class SW(object):
    """docstring for SW"""
    def __init__(self, account, password,url):
        super(SW, self).__init__()
        self.url = url
        self.account = account
        self.password = password
        self.session = self.login()
    
    HEADERS = {
        "User-Agent":"Mozilla/5.0 (Linux; U; Mobile; Android 6.0.1;C107-9 Build/FRF91 )",
        "Referer": "http://www.baidu.com",
        "Accept-encoding": "gzip, deflate, br",
        "Accept-language": "zh-CN,zh-TW;q=0.8,zh;q=0.6,en;q=0.4,ja;q=0.2",
        "Cache-control": "max-age=0"
    }

    def login(self):
        params = {
            "method" : "authUser",
            "xh" : self.account,
            "pwd" : self.password
        }
        session = requests.Session()
        req = session.get(self.url, params = params, timeout = 5, headers = self.HEADERS)
        s = json.loads(req.text)
        print(s)
        if s["flag"] != "1" : exit(0)
        self.HEADERS["token"] = s["token"]
        return session

    

    def get_handle(self,params):
        req = self.session.get(self.url, params = params ,timeout = 5 ,headers = self.HEADERS)
        return req

    def get_student_info(self):
        params = {
            "method" : "getUserInfo",
            "xh" : self.account
        }
        req = self.get_handle(params)
        print(req.text)
    
    def get_current_time(self):
        params = {
            "method" : "getCurrentTime",
            "currDate" : datetime.datetime.now().strftime("%Y-%m-%d")
        }
        req = self.get_handle(params)
        print(req.text)
        return req.text

    def get_class_info(self,zc = -1):
        s = json.loads(self.get_current_time())
        params={
            "method" : "getKbcxAzc",
            "xnxqid" : s["xnxqh"],
            "zc" : s["zc"] if zc == -1 else zc,
            "xh" : self.account
        }
        req = self.get_handle(params)
        print(req.text)

    def get_classroom_info(self,idleTime = "allday"):
        params={
            "method" : "getKxJscx",
            "time" : datetime.datetime.now().strftime("%Y-%m-%d"),
            "idleTime" : idleTime
        }
        req = self.get_handle(params)
        print(req.text)

    def get_grade_info(self,sy = ""):
        params={
            "method" : "getCjcx",
            "xh" : self.account,
            "xnxqid" : sy
        }
        req = self.get_handle(params)
        print("全部成绩" if sy == "" else sy)
        s = json.loads(req.text)
        if s[0] != None :
            s = json.loads(req.text)
            for x in s:
                print("%s   %d   %s" % (str(x["zcj"]),x["xf"],x["kcmc"]))
            print("绩点：" + str(self.get_gp(s)))
        else :
            print("空")

    def get_exam_info(self):
        params={
            "method" : "getKscx",
            "xh" : self.account,
        }
        req = self.get_handle(params)
        print(req.text)

    def get_gp(self,data):
        GP = 0.0
        credit = 0.0
        for x in data:
            credit += x["xf"]
            if  x["zcj"] == "优":
                GP += (4.5 * x["xf"])
            elif x["zcj"] == "良":
                GP += (3.5 * x["xf"])
            elif x["zcj"] == "中":
                GP += (2.5 * x["xf"])
            elif x["zcj"] == "及格":
                GP += (1.5 * x["xf"])
            elif x["zcj"] == "不及格":
                GP += 0 
            else :
                if float(x["zcj"]) >= 60:
                   GP += (((float(x["zcj"])-60)/10+1) * x["xf"])
        return GP/credit


if __name__ == "__main__":
    Q = SW(account,password,url)
    # Q.get_student_info() #获取学生信息
    # Q.get_current_time() #获取学年信息
    # Q.get_class_info() #当前周次课表
    # Q.get_class_info(3) #指定周次课表
    # Q.get_classroom_info("0102") #空教室查询 "allday"：全天 "am"：上午 "pm"：下午 "night"：晚上 "0102":1.2节空教室 "0304":3.4节空教室
    # Q.get_grade_info("2018-2019-1") #成绩查询 #无参数查询全部成绩
    # Q.get_exam_info() #获取考试信息


